Bug/69450 complete specs for autocompleter changes

dombesz opened the pull request:
# Ticket
https://community.openproject.org/wp/69450

# What are you trying to accomplish?

## Screenshots
<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->

# What approach did you choose and why?
- [X] Add specs for workspace badges in header dropdown, project settings, and WP table inline edit.
  - Add spec helper classes for the select inputs. The ProjectEditField is for the project selectors, InlineProjectEditField is for project selectors for the work package table inline editing.
- [X] Refactor icon rendering.
  - Imperatively add the icon in the template instead of programatically rendering it from the component.
  - Use the angular `@switch` block in templates instead of a component method.
- [X] Selected value now shows badges.
  - Fix badge rendering of selected values in multi select mode. 
  <img width="338" height="84" alt="image" src="https://github.com/user-attachments/assets/2f38539e-559f-41d4-858b-bd80f443d6c5" />
  
  - Fix filter list expanding over the viewport when the selected project name is long enough to leave the viewport.
  <img width="206" height="146" alt="image" src="https://github.com/user-attachments/assets/ab4e9d66-be06-4118-ba5c-69e488b8c0b7" /> <details><summary>Before and after of ellipsis not working due to the container list expanding over the viewport.</summary><b> Before</b><p> <img width="952" height="681" alt="image" src="https://github.com/user-attachments/assets/9f56b1c1-3e5c-4852-b69a-29151901ae4e" /> </p><b> After</b><p><img width="960" height="735" alt="image" src="https://github.com/user-attachments/assets/6ed1e31c-a13a-4f92-9004-f311065aa9b8" /></p></details>
  
- [X] Fix accesibility of the remove icon on the selected options in multi select mode.
  - Close icon (×) now functions as a button, can be focused and triggered by pressing enter.
  <img width="312" height="92" alt="image" src="https://github.com/user-attachments/assets/c8a08821-2a04-4a8d-a17d-c8862b28449c" />
- [X] Primer rspec selector fixes, `:octicon` selector now uses CSS instead of XPath, `:primer_text` handles nil text and color combination.

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

dombesz committed:
Add Program and Portfolio icon to the selected value as well.

This reverts commit c732e3498ffef6b190c6a51fbcda5676fc62c59a.

dombesz committed:
[#69450] Complete specs for autocompleter changes

https://community.openproject.org/work_packages/69450

myabc commented on frontend/src/app/shared/components/autocompleter/project-autocompleter/project-autocompleter-template.component.html:
```diff
@@ -4,7 +4,16 @@
   >
   <span class="ng-value-icon left" (click)="clear(item)">×</span>
-  <span class="ng-value-label">{{ item.name }}</span>
+  <span class="ng-value-label" style="display: flex; gap: 0.5rem;">
```
We can use utility classes here: [`.d-flex` from Primer](`https://primer.style/product/css-utilities/flexbox/`) and our own (`.gap-*`)[src/global_styles/openproject/_gap.scss} introduced by @toy 

```suggestion
  <span class="ng-value-label" class="d-flex gap-3">
```

myabc commented on frontend/src/app/shared/components/autocompleter/project-autocompleter/project-autocompleter-template.component.html:
```diff
@@ -4,7 +4,16 @@
+      <span
+        class="color-fg-muted"
+        [innerHTML]="workspaceTypeIconWithLabel(item)"
+        style="flex-shrink: 0;"
```
```suggestion
        class="color-fg-muted flex-shrink-0"
        [innerHTML]="workspaceTypeIconWithLabel(item)"
```

myabc commented on spec/features/projects/project_autocomplete_spec.rb:
```diff
@@ -187,4 +184,33 @@
+    top_menu.within_item(portfolio) do
+      expect(page).to have_text("Test Portfolio")
+      expect(page).to have_css("svg.octicon")
+    end
```
We have some specific Capybara selectors/RSpec matchers just for this purpose - see `spec/support/capybara/primer_component_selectors.rb`,

myabc replied:
I've added some suggestions inline. The Capybara selectors are more geared to component specs, but should still work here. Please feel free to ping me or @cbliard if you run into any issues.

myabc commented on spec/support/edit_fields/project_edit_field.rb:
```diff
@@ -0,0 +1,45 @@
+
+      if workspace_badge
+        expect(option).to have_css("svg.octicon")
+        expect(option).to have_css(".color-fg-muted", text: workspace_badge)
```
```suggestion
        expect(option).to have_octicon # optionally specif name
        expect(option).to have_primer_text(workspace_badge, color: :muted)
```

cbliard replied:
Nice to use our helpers here 👍 

This could be suggested by custom rubocop cops.
That would relieve you from having to monitor all PRs about it, and allow faster feedback for people writing this kind of code.
What do you think?

myabc commented on spec/support/edit_fields/project_edit_field.rb:
```diff
@@ -0,0 +1,45 @@
+      else
+        # Expect no octicon SVG
+        expect(option).to have_no_css("svg.octicon")
+        expect(option).to have_no_css(".color-fg-muted")
```
```suggestion
        expect(option).to have_no_octicon # optionally specif name
        expect(option).to have_no_primer_text
```

myabc commented on frontend/src/app/shared/components/autocompleter/project-autocompleter/project-autocompleter-template.component.html:
```diff
@@ -4,7 +4,16 @@
-  <span class="ng-value-label">{{ item.name }}</span>
+  <span class="ng-value-label" style="display: flex; gap: 0.5rem;">
+    <span class="ellipsis">{{ item.name }}</span>
+    @if (shouldShowWorkspaceTypeBadge(item)) {
```
I have to admit that I'm not a fan of how this was implemented in #21223.

Why not just use a simple [`@switch`](https://angular.dev/guide/templates/control-flow#conditionally-display-content-with-the-switch-block) in the template?

```html
@switch (item._type) {
  @case ('Portfolio') {
    <svg version-icon size="small" />
  }
  @case ('Program') {
    <svg briefcase-icon size="small" />
  }
}
```

myabc reviewed (commented):
Sorry for chiming in quite late on this particular functionality. However I think it might make sense to try to align styling closer to Primer - in particular, we should look at the ActionList component:

[Action List inline descriptions](https://primer.style/product/components/action-list/#items-with-inline-descriptions) would probably be the closest mapping. The missing CSS properties are as follows:

```
line-height: 16px;
font-size: var(--text-body-size-small, .75rem);
font-weight: var(--base-text-weight-normal, 400);
```

Alternatively there could be an argument for styling as [trailing content](https://primer.style/product/components/action-list/#items-with-leading-and-trailing-content). However it's probably worth going through the [guidelines first](https://primer.style/product/components/action-list/guidelines/#trailing-visual-and-trailing-text).

dombesz committed:
Use the have_octicon and have_primer_text_field custom matchers. Maintain the gap for multi select project autocompleters.

dombesz committed:
Use a switch block in the template to decide which icon to display, use icon tags instead of programatically rendering them.

dombesz committed:
Use switch statement and render svg icon tags in the header project selector.

dombesz committed:
Create a separate project edit helper for the work packages table

dombesz committed:
Make the multi select selected option close button accessible.

dombesz committed:
Extend top menu expect result method to handle workspace badges

dombesz committed:
Skip checking for badge when it is not provided to the expectation.

dombesz committed:
Fix filter list item expanding over the viewport.

dombesz commented:
> Sorry for chiming in quite late on this particular functionality. However I think it might make sense to try to align styling closer to Primer - in particular, we should look at the ActionList component:
> 
> [Action List inline descriptions](https://primer.style/product/components/action-list/#items-with-inline-descriptions) would probably be the closest mapping. The missing CSS properties are as follows:
> 
> ```
> line-height: 16px;
> font-size: var(--text-body-size-small, .75rem);
> font-weight: var(--base-text-weight-normal, 400);
> ```
> 
> Alternatively there could be an argument for styling as [trailing content](https://primer.style/product/components/action-list/#items-with-leading-and-trailing-content). However it's probably worth going through the [guidelines first](https://primer.style/product/components/action-list/guidelines/#trailing-visual-and-trailing-text).

Thank you for the thorough review @myabc ! I addressed all your comments, except the one for aligning the label more to primer. I would like to do that separately as that involves the product team as well.

dombesz committed:
Increase gap between project name and icon

dombesz committed:
Fix ellipsis on selected projects with badge

dombesz committed:
Fix icon mixup between project and portfolio.

ulferts reviewed (approved):
As discussed, I only checked the specs in depth trusting on @myabc to focus on the productive part.

myabc commented on spec/support/capybara/primer_component_selectors.rb:
```diff
@@ -101,11 +104,10 @@
 
-  xpath do |locator|
-    xpath = XPath.descendant(:svg)
-    xpath = builder(xpath).add_attribute_conditions(class: "octicon")
-    xpath = builder(xpath).add_attribute_conditions(class: "octicon-#{locator.to_s.downcase}") if locator
-    xpath
+  css do |locator|
+    css_selector = "svg.octicon"
+    css_selector += ".octicon-#{locator.to_s.downcase}" if locator
+    css_selector
   end
```
@dombesz why this change?

dombesz replied:
For some reason the capybara rendered html did not work the same way with the xpath selector, but it does with the css one. (I assume javascript changes some elements in the browser and the xpath doesn't match up anymore, the css selector is a bit more robust)

myabc replied:
Interesting. I've only really used this selector in component specs (these selectors are more geared to testing basic rendering).

myabc commented on frontend/src/app/shared/components/autocompleter/project-autocompleter/project-autocompleter-template.component.html:
```diff
@@ -3,8 +3,32 @@
   let-clear="clear"
   >
-  <span class="ng-value-icon left" (click)="clear(item)">×</span>
-  <span class="ng-value-label">{{ item.name }}</span>
+  <span class="d-flex">
+    <span
+      class="ng-value-icon left"
+      tabindex="0"
+      role="button"
+      (click)="clear(item)"
+      (keydown.enter)="clear(item)"
+    >×</span>
```
@dombesz thanks for fixing the accessibility error!

myabc commented on frontend/src/app/shared/components/autocompleter/project-autocompleter/project-autocompleter-template.component.html:
```diff
@@ -16,20 +40,27 @@
-          style="flex-shrink: 0;"
-        ></div>
+      >{{ item.name }}</span>
+      @switch (item._type) {
```
Ditto. Check `@if (portfolioModelsEnabled)`?

dombesz replied:
Thanks for catching this. Fixed.

myabc commented on frontend/src/app/shared/components/autocompleter/project-autocompleter/project-autocompleter-template.component.html:
```diff
@@ -3,8 +3,32 @@
+    >×</span>
+    <span class="ng-value-label d-flex gap-2">
+      <span class="ellipsis">{{ item.name }}</span>
+      @switch (item._type) {
```
Is there a reason we don't check `@if (portfolioModelsEnabled)` here?

dombesz replied:
Thanks for catching this. Fixed.

myabc commented on frontend/src/global_styles/content/_autocomplete.sass:
```diff
@@ -140,9 +140,12 @@ div.autocomplete
   min-height: initial !important
 
-  .ng-value-label
+  .ng-value-label:not(.d-flex)
     display: initial !important
 
+  .ng-value-label.d-flex
+    min-width: 0
```
nitpick: even though I'm guilty of doing it at times, I don't really like mixing utility and component/semantic classes like this. 

also, is this actually needed?

dombesz replied:
Yes, they are important.
<details>
  <summary>The first one fixes long named items in the work package multi select. (Click for screenshot)</summary>
  <p>
    <img width="616" height="250" alt="image" src="https://github.com/user-attachments/assets/05dc13cd-bc54-4b5d-8e4c-7d1bcfa4ecc4" />
  </p>
  </details>

<details>
  <summary>The second ensures the badge is not pushed out of the item box. (Click for screenshot)</summary>
  <p>
   <img width="330" height="124" alt="image" src="https://github.com/user-attachments/assets/d2a2f5b2-f292-4216-80f0-5ccb721b4bc9" />

  </p>
  </details>

 
I changed the selectors d87a99bc0c05c951640f36dcdedc9d9eb0710f82 to decide on styling whether there is a child tag or not.

myabc reviewed (commented):
Hi @dombesz. Sorry for the delay in reviewing this again. This looks really good - thanks for making these changes.

I just have a couple small suggestions, in addition to the inline comments:

<img width="473" height="229" alt="Image" src="https://github.com/user-attachments/assets/59d41046-dfae-4a09-a992-80cc013e7885" />

In the Header Project selector, we are showing both the project name and project type in bold when filtered. IMO we should only show the project name in bold.

I'd personally advocate for showing the project type as a badge ([Primer Label](https://primer.style/product/getting-started/rails/components/label/)) but this could be addressed in a follow up PR. I'd also want a second opinion from @HDinger or a designer before pushing for it!

dombesz committed:
Inlcude space key and aria-label in the accessiblity implementation.

dombesz committed:
Add missing the feature flag guard

dombesz committed:
Do not mix utility and classes, decide on styling whether the element has any children.

dombesz commented:
Thanks @myabc , I addressed your comments, except the following:

> In the Header Project selector, we are showing both the project name and project type in bold when filtered. IMO we should only show the project name in bold.

This one is already addressed in [a follow up PR](https://github.com/opf/openproject/pull/21364), which also applies the smaller text size for the badge as you suggested in [a previous comment](https://github.com/opf/openproject/pull/21261#issuecomment-3616548406).

As per the Primer label, I agree that it should be discussed first.

myabc reviewed (approved):
Looks good! Thanks for all the changes
